In [1]:
import pandas as pd

# orderdetailsdf = pd.read_excel(r'Order Details.xlsx')
ordersdf = pd.read_excel(r'Orders.xlsx')

ordersdf.head()

Order ID                   Customer           Employee Order Date  \
0     10248  Vins et alcools Chevalier   Buchanan, Steven 1994-08-04   
1     10249         Toms Spezialitäten    Suyama, Michael 1994-08-05   
2     10250              Hanari Carnes  Peacock, Margaret 1994-08-08   
3     10251       Victuailles en stock   Leverling, Janet 1994-08-08   
4     10252           Suprêmes délices  Peacock, Margaret 1994-08-09   

  Required Date Shipped Date          Ship Via  Freight  \
0    1994-09-01   1994-08-16  Federal Shipping    32.38   
1    1994-09-16   1994-08-10    Speedy Express    11.61   
2    1994-09-05   1994-08-12    United Package    65.83   
3    1994-09-05   1994-08-15    Speedy Express    41.34   
4    1994-09-06   1994-08-11    United Package    51.30   

                   Ship Name          Ship Address       Ship City  \
0  Vins et alcools Chevalier    59 rue de l'Abbaye           Reims   
1         Toms Spezialitäten         Luisenstr. 48         Münster   
2              Hanari Carnes       Rua do Paço, 67  Rio de Janeiro   
3       Victuailles en stock    2, rue du Commerce            Lyon   
4           Suprêmes délices  Boulevard Tirou, 255       Charleroi   

  Ship Region Ship Postal Code Ship Country  
0         NaN            51100       France  
1         NaN            44087      Germany  
2          RJ        05454-876       Brazil  
3         NaN            69004       France  
4         NaN           B-6000      Belgium

In [2]:
ordersdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 830 entries, 0 to 829
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Order ID          830 non-null    int64         
 1   Customer          830 non-null    object        
 2   Employee          830 non-null    object        
 3   Order Date        830 non-null    datetime64[ns]
 4   Required Date     830 non-null    datetime64[ns]
 5   Shipped Date      809 non-null    datetime64[ns]
 6   Ship Via          830 non-null    object        
 7   Freight           830 non-null    float64       
 8   Ship Name         830 non-null    object        
 9   Ship Address      830 non-null    object        
 10  Ship City         830 non-null    object        
 11  Ship Region       323 non-null    object        
 12  Ship Postal Code  811 non-null    object        
 13  Ship Country      830 non-null    object        
dtypes: datetime64[ns](3), floa

In [3]:
ordersdf.dropna(subset = ["Shipped Date"], inplace=True)

In [4]:
ordersdf

Order ID                      Customer           Employee Order Date  \
0       10248     Vins et alcools Chevalier   Buchanan, Steven 1994-08-04   
1       10249            Toms Spezialitäten    Suyama, Michael 1994-08-05   
2       10250                 Hanari Carnes  Peacock, Margaret 1994-08-08   
3       10251          Victuailles en stock   Leverling, Janet 1994-08-08   
4       10252              Suprêmes délices  Peacock, Margaret 1994-08-09   
..        ...                           ...                ...        ...   
815     11063  Hungry Owl All-Night Grocers   Leverling, Janet 1996-05-30   
816     11064            Save-a-lot Markets     Davolio, Nancy 1996-05-31   
818     11066          White Clover Markets       King, Robert 1996-05-31   
819     11067      Drachenblut Delikatessen     Davolio, Nancy 1996-06-03   
821     11069           Tortuga Restaurante     Davolio, Nancy 1996-06-03   

    Required Date Shipped Date          Ship Via  Freight  \
0      1994-09-01   1994-08-16  Federal Shipping    32.38   
1      1994-09-16   1994-08-10    Speedy Express    11.61   
2      1994-09-05   1994-08-12    United Package    65.83   
3      1994-09-05   1994-08-15    Speedy Express    41.34   
4      1994-09-06   1994-08-11    United Package    51.30   
..            ...          ...               ...      ...   
815    1996-06-27   1996-06-05    United Package    81.73   
816    1996-06-28   1996-06-03    Speedy Express    30.09   
818    1996-06-28   1996-06-03    United Package    44.72   
819    1996-06-17   1996-06-05    United Package     7.98   
821    1996-07-01   1996-06-05    United Package    15.67   

                        Ship Name          Ship Address       Ship City  \
0       Vins et alcools Chevalier    59 rue de l'Abbaye           Reims   
1              Toms Spezialitäten         Luisenstr. 48         Münster   
2                   Hanari Carnes       Rua do Paço, 67  Rio de Janeiro   
3            Victuailles en stock    2, rue du Commerce            Lyon   
4                Suprêmes délices  Boulevard Tirou, 255       Charleroi   
..                            ...                   ...             ...   
815  Hungry Owl All-Night Grocers      8 Johnstown Road            Cork   
816            Save-a-lot Markets       187 Suffolk Ln.           Boise   
818          White Clover Markets   1029 - 12th Ave. S.         Seattle   
819      Drachenblut Delikatessen          Walserweg 21          Aachen   
821           Tortuga Restaurante      Avda. Azteca 123     México D.F.   

    Ship Region Ship Postal Code Ship Country  
0           NaN            51100       France  
1           NaN            44087      Germany  
2            RJ        05454-876       Brazil  
3           NaN            69004       France  
4           NaN           B-6000      Belgium  
..          ...              ...          ...  
815    Co. Cork              NaN      Ireland  
816          ID            83720          USA  
818          WA            98124          USA  
819         NaN            52066      Germany  
821         NaN            05033       Mexico  

[809 rows x 14 columns]

In [5]:
import numpy as np
ordersdf['diff_days'] = ordersdf['Shipped Date'] - ordersdf['Order Date']

ordersdf['diff_days'] = ordersdf['diff_days'] / np.timedelta64(1,'D')

In [6]:
# ordersdf['Order Date'] = ordersdf['Order Date'].dt.strftime('%Y')
# ordersdf['Shipped Date'] = ordersdf['Shipped Date'].dt.strftime('%Y')

In [7]:
ordersdf

Order ID                      Customer           Employee Order Date  \
0       10248     Vins et alcools Chevalier   Buchanan, Steven 1994-08-04   
1       10249            Toms Spezialitäten    Suyama, Michael 1994-08-05   
2       10250                 Hanari Carnes  Peacock, Margaret 1994-08-08   
3       10251          Victuailles en stock   Leverling, Janet 1994-08-08   
4       10252              Suprêmes délices  Peacock, Margaret 1994-08-09   
..        ...                           ...                ...        ...   
815     11063  Hungry Owl All-Night Grocers   Leverling, Janet 1996-05-30   
816     11064            Save-a-lot Markets     Davolio, Nancy 1996-05-31   
818     11066          White Clover Markets       King, Robert 1996-05-31   
819     11067      Drachenblut Delikatessen     Davolio, Nancy 1996-06-03   
821     11069           Tortuga Restaurante     Davolio, Nancy 1996-06-03   

    Required Date Shipped Date          Ship Via  Freight  \
0      1994-09-01   1994-08-16  Federal Shipping    32.38   
1      1994-09-16   1994-08-10    Speedy Express    11.61   
2      1994-09-05   1994-08-12    United Package    65.83   
3      1994-09-05   1994-08-15    Speedy Express    41.34   
4      1994-09-06   1994-08-11    United Package    51.30   
..            ...          ...               ...      ...   
815    1996-06-27   1996-06-05    United Package    81.73   
816    1996-06-28   1996-06-03    Speedy Express    30.09   
818    1996-06-28   1996-06-03    United Package    44.72   
819    1996-06-17   1996-06-05    United Package     7.98   
821    1996-07-01   1996-06-05    United Package    15.67   

                        Ship Name          Ship Address       Ship City  \
0       Vins et alcools Chevalier    59 rue de l'Abbaye           Reims   
1              Toms Spezialitäten         Luisenstr. 48         Münster   
2                   Hanari Carnes       Rua do Paço, 67  Rio de Janeiro   
3            Victuailles en stock    2, rue du Commerce            Lyon   
4                Suprêmes délices  Boulevard Tirou, 255       Charleroi   
..                            ...                   ...             ...   
815  Hungry Owl All-Night Grocers      8 Johnstown Road            Cork   
816            Save-a-lot Markets       187 Suffolk Ln.           Boise   
818          White Clover Markets   1029 - 12th Ave. S.         Seattle   
819      Drachenblut Delikatessen          Walserweg 21          Aachen   
821           Tortuga Restaurante      Avda. Azteca 123     México D.F.   

    Ship Region Ship Postal Code Ship Country  diff_days  
0           NaN            51100       France       12.0  
1           NaN            44087      Germany        5.0  
2            RJ        05454-876       Brazil        4.0  
3           NaN            69004       France        7.0  
4           NaN           B-6000      Belgium        2.0  
..          ...              ...          ...        ...  
815    Co. Cork              NaN      Ireland        6.0  
816          ID            83720          USA        3.0  
818          WA            98124          USA        3.0  
819         NaN            52066      Germany        2.0  
821         NaN            05033       Mexico        2.0  

[809 rows x 15 columns]

In [8]:
ordersdf2 = ordersdf.groupby(by=['Ship Via'], sort=False, as_index=False)['diff_days'].sum()
ordersdf2

Ship Via  diff_days
0  Federal Shipping     1861.0
1    Speedy Express     2100.0
2    United Package     2909.0

In [9]:
ordersdf[ordersdf['Ship Via'] == 'Federal Shipping']

Order ID                    Customer           Employee Order Date  \
0       10248   Vins et alcools Chevalier   Buchanan, Steven 1994-08-04   
7       10255          Richter Supermarkt    Dodsworth, Anne 1994-08-12   
9       10257            HILARIÓN-Abastos  Peacock, Margaret 1994-08-16   
11      10259  Centro comercial Moctezuma  Peacock, Margaret 1994-08-18   
14      10262  Rattlesnake Canyon Grocery    Callahan, Laura 1994-08-22   
..        ...                         ...                ...        ...   
785     11033          Richter Supermarkt       King, Robert 1996-05-17   
788     11036    Drachenblut Delikatessen    Callahan, Laura 1996-05-20   
799     11047          Eastern Connection       King, Robert 1996-05-24   
800     11048       Bottom-Dollar Markets       King, Robert 1996-05-24   
809     11057                 North/South   Leverling, Janet 1996-05-29   

    Required Date Shipped Date          Ship Via  Freight  \
0      1994-09-01   1994-08-16  Federal Shipping    32.38   
7      1994-09-09   1994-08-15  Federal Shipping   148.33   
9      1994-09-13   1994-08-22  Federal Shipping    81.91   
11     1994-09-15   1994-08-25  Federal Shipping     3.25   
14     1994-09-19   1994-08-25  Federal Shipping    48.29   
..            ...          ...               ...      ...   
785    1996-06-14   1996-05-23  Federal Shipping    84.74   
788    1996-06-17   1996-05-22  Federal Shipping   149.47   
799    1996-06-21   1996-05-31  Federal Shipping    46.62   
800    1996-06-21   1996-05-30  Federal Shipping    24.12   
809    1996-06-26   1996-05-31  Federal Shipping     4.13   

                      Ship Name                                Ship Address  \
0     Vins et alcools Chevalier                          59 rue de l'Abbaye   
7            Richter Supermarkt                                 Starenweg 5   
9              HILARIÓN-Abastos  Carrera 22 con Ave. Carlos Soublette #8-35   
11   Centro comercial Moctezuma                     Sierras de Granada 9993   
14   Rattlesnake Canyon Grocery                             2817 Milton Dr.   
..                          ...                                         ...   
785          Richter Supermarkt                                 Starenweg 5   
788    Drachenblut Delikatessen                                Walserweg 21   
799          Eastern Connection                              35 King George   
800       Bottom-Dollar Markets                          23 Tsawassen Blvd.   
809                 North/South               South House\n300 Queensbridge   

         Ship City Ship Region Ship Postal Code Ship Country  diff_days  
0            Reims         NaN            51100       France       12.0  
7           Genève         NaN             1204  Switzerland        3.0  
9    San Cristóbal     Táchira             5022    Venezuela        6.0  
11     México D.F.         NaN            05022       Mexico        7.0  
14     Albuquerque          NM            87110          USA        3.0  
..             ...         ...              ...          ...        ...  
785         Genève         NaN             1204  Switzerland        6.0  
788         Aachen         NaN            52066      Germany        2.0  
799         London         NaN          WX3 6FW           UK        7.0  
800      Tsawassen          BC          T2F 8M4       Canada        6.0  
809         London         NaN          SW7 1RZ           UK        2.0  

[249 rows x 15 columns]

In [10]:
Average = 1861.0/249
Average

7.473895582329317

In [11]:
ordersdf[ordersdf['Ship Via'] == 'Speedy Express']

Order ID              Customer           Employee Order Date  \
1       10249    Toms Spezialitäten    Suyama, Michael 1994-08-05   
3       10251  Victuailles en stock   Leverling, Janet 1994-08-08   
10      10258          Ernst Handel     Davolio, Nancy 1994-08-17   
12      10260    Ottilies Käseladen  Peacock, Margaret 1994-08-19   
17      10265  Blondel père et fils     Fuller, Andrew 1994-08-25   
..        ...                   ...                ...        ...   
794     11042      Comércio Mineiro     Fuller, Andrew 1996-05-22   
796     11044        Wolski  Zajazd  Peacock, Margaret 1996-05-23   
801     11049   Gourmet Lanchonetes   Leverling, Janet 1996-05-24   
804     11052         Hanari Carnes   Leverling, Janet 1996-05-27   
816     11064    Save-a-lot Markets     Davolio, Nancy 1996-05-31   

    Required Date Shipped Date        Ship Via  Freight             Ship Name  \
1      1994-09-16   1994-08-10  Speedy Express    11.61    Toms Spezialitäten   
3      1994-09-05   1994-08-15  Speedy Express    41.34  Victuailles en stock   
10     1994-09-14   1994-08-23  Speedy Express   140.51          Ernst Handel   
12     1994-09-16   1994-08-29  Speedy Express    55.09    Ottilies Käseladen   
17     1994-09-22   1994-09-12  Speedy Express    55.28  Blondel père et fils   
..            ...          ...             ...      ...                   ...   
794    1996-06-05   1996-05-31  Speedy Express    29.99      Comércio Mineiro   
796    1996-06-20   1996-05-31  Speedy Express     8.72         Wolski Zajazd   
801    1996-06-21   1996-06-03  Speedy Express     8.34   Gourmet Lanchonetes   
804    1996-06-24   1996-05-31  Speedy Express    67.26         Hanari Carnes   
816    1996-06-28   1996-06-03  Speedy Express    30.09    Save-a-lot Markets   

             Ship Address       Ship City Ship Region Ship Postal Code  \
1           Luisenstr. 48         Münster         NaN            44087   
3      2, rue du Commerce            Lyon         NaN            69004   
10           Kirchgasse 6            Graz         NaN             8010   
12     Mehrheimerstr. 369            Köln         NaN            50739   
17       24, place Kléber      Strasbourg         NaN            67000   
..                    ...             ...         ...              ...   
794  Av. dos Lusíadas, 23       São Paulo          SP        05432-043   
796       ul. Filtrowa 68        Warszawa         NaN           01-012   
801       Av. Brasil, 442        Campinas          SP        04876-786   
804       Rua do Paço, 67  Rio de Janeiro          RJ        05454-876   
816       187 Suffolk Ln.           Boise          ID            83720   

    Ship Country  diff_days  
1        Germany        5.0  
3         France        7.0  
10       Austria        6.0  
12       Germany       10.0  
17        France       18.0  
..           ...        ...  
794       Brazil        9.0  
796       Poland        8.0  
801       Brazil       10.0  
804       Brazil        4.0  
816          USA        3.0  

[245 rows x 15 columns]

In [12]:
Average = 2100.0/245
Average

8.571428571428571

In [13]:
ordersdf[ordersdf['Ship Via'] == 'United Package']

Order ID                      Customer           Employee Order Date  \
2       10250                 Hanari Carnes  Peacock, Margaret 1994-08-08   
4       10252              Suprêmes délices  Peacock, Margaret 1994-08-09   
5       10253                 Hanari Carnes   Leverling, Janet 1994-08-10   
6       10254             Chop-suey Chinese   Buchanan, Steven 1994-08-11   
8       10256        Wellington Importadora   Leverling, Janet 1994-08-15   
..        ...                           ...                ...        ...   
812     11060                Franchi S.p.A.     Fuller, Andrew 1996-05-30   
815     11063  Hungry Owl All-Night Grocers   Leverling, Janet 1996-05-30   
818     11066          White Clover Markets       King, Robert 1996-05-31   
819     11067      Drachenblut Delikatessen     Davolio, Nancy 1996-06-03   
821     11069           Tortuga Restaurante     Davolio, Nancy 1996-06-03   

    Required Date Shipped Date        Ship Via  Freight  \
2      1994-09-05   1994-08-12  United Package    65.83   
4      1994-09-06   1994-08-11  United Package    51.30   
5      1994-08-24   1994-08-16  United Package    58.17   
6      1994-09-08   1994-08-23  United Package    22.98   
8      1994-09-12   1994-08-17  United Package    13.97   
..            ...          ...             ...      ...   
812    1996-06-27   1996-06-03  United Package    10.98   
815    1996-06-27   1996-06-05  United Package    81.73   
818    1996-06-28   1996-06-03  United Package    44.72   
819    1996-06-17   1996-06-05  United Package     7.98   
821    1996-07-01   1996-06-05  United Package    15.67   

                        Ship Name          Ship Address       Ship City  \
2                   Hanari Carnes       Rua do Paço, 67  Rio de Janeiro   
4                Suprêmes délices  Boulevard Tirou, 255       Charleroi   
5                   Hanari Carnes       Rua do Paço, 67  Rio de Janeiro   
6               Chop-suey Chinese          Hauptstr. 31            Bern   
8          Wellington Importadora    Rua do Mercado, 12         Resende   
..                            ...                   ...             ...   
812                Franchi S.p.A.   Via Monte Bianco 34          Torino   
815  Hungry Owl All-Night Grocers      8 Johnstown Road            Cork   
818          White Clover Markets   1029 - 12th Ave. S.         Seattle   
819      Drachenblut Delikatessen          Walserweg 21          Aachen   
821           Tortuga Restaurante      Avda. Azteca 123     México D.F.   

    Ship Region Ship Postal Code Ship Country  diff_days  
2            RJ        05454-876       Brazil        4.0  
4           NaN           B-6000      Belgium        2.0  
5            RJ        05454-876       Brazil        6.0  
6           NaN             3012  Switzerland       12.0  
8            SP        08737-363       Brazil        2.0  
..          ...              ...          ...        ...  
812         NaN            10100        Italy        4.0  
815    Co. Cork              NaN      Ireland        6.0  
818          WA            98124          USA        3.0  
819         NaN            52066      Germany        2.0  
821         NaN            05033       Mexico        2.0  

[315 rows x 15 columns]

In [14]:
Average = 2909.0/315
Average

9.234920634920634

### Rata-rata waktu tunggu barang dari permintaan (Order) sampai dikirimkan (Shipped):
### Federal Shipping: 7.47 hari.
### Speedy Express: 8.57 hari.
### United Package: 9.23 hari.

### Dapat dilihat bahwa Federal Shipping dapat merespon paket dengan lebih cepat.